packages/stack/proxy/src/rpc-manager.ts

```diff
@@ -0,0 +1,193 @@
+import Web3 from "web3";
+
+export type RpcRequestInterceptor<T> = (params: ProxyRequestParams<T>) => ProxyRequestParams<T>;
+export type RpcResponseInterceptor<T, R> = (params: ProxyRequestParams<T> | ProxyResponseParams<T, R>) => ProxyResponseParams<T, R>;
+
+interface RegistrationOptions {
+  priority: number;
+}
+
+interface RpcRegistration {
+  filter: string | string[] | RegExp;
+  options: RegistrationOptions;
+}
+
+interface RpcRequestRegistration<T> extends RpcRegistration {
+  interceptor: RpcRequestInterceptor<T>;
+}
+
+interface RpcResponseRegistration<T, R> extends RpcRegistration {
+  interceptor: RpcResponseInterceptor<T, R>;
+}
+
+export default class RpcManager {
+
+  private _web3: Web3 | null = null;
+
+  private logger: Logger;
+
+  private events: EmbarkEvents;
+
+  public _accounts: Account[] | null = null;
+
+  public _nodeAccounts: string[] | null = null;
+
+  private requestInterceptors: Array<RpcRequestRegistration<any>> = [];
+
+  private responseInterceptors: Array<RpcResponseRegistration<any, any>> = [];
+
+  constructor(private readonly embark: Embark) {
+    this.events = embark.events;
+    this.logger = embark.logger;
+  }
+
+  protected get web3() {
+    return (async () => {
+      if (!this._web3) {
+        await this.events.request2("blockchain:started");
+        // get connection directly to the node
+        const provider = await this.events.request2("blockchain:node:provider", "ethereum");
+        this._web3 = new Web3(provider);
+      }
+      return this._web3;
+    })();
+  }
+
+  private get nodeAccounts() {
+    return (async () => {
+      if (!this._nodeAccounts) {
+        const web3 = await this.web3;
+        this._nodeAccounts = await web3.eth.getAccounts();
+      }
+      return this._nodeAccounts || [];
+    })();
+  }
+
+  private get accounts() {
+    return (async () => {
+      if (!this._accounts) {
+        const web3 = await this.web3;
+        const nodeAccounts = await this.nodeAccounts;
+        this._accounts = AccountParser.parseAccountsConfig(this.embark.config.blockchainConfig.accounts, web3, dappPath(), this.logger, nodeAccounts);
+      }
+      return this._accounts || [];
+    })();
+  }
+
+  public init() {
+    this.registerActions();
+    this.setCommandHandlers();
+  }
+
+  private registerActions() {
+    this.embark.registerActionForEvent("tests:config:updated", { priority: 40 }, (_params, cb) => {
+      // blockchain configs may have changed (ie endpoint)
+      this._web3 = null;
+
+      // web.eth.getAccounts may return a different value now
+      // update accounts across all modifiers
+      this.resetAccounts(cb);
+    });
+
+    this.embark.registerActionForEvent("blockchain:proxy:request", this.onProxyRequest.bind(this));
+    this.embark.registerActionForEvent("blockchain:proxy:response", this.onProxyResponse.bind(this));
+  }
+
+  private setCommandHandlers() {
+    this.events.setCommandHandler("rpc:accounts:reset", this.resetAccounts.bind(this));
```

Comment:
This will also not be needed here and should be moved to the `RpcInterceptor` base class, along with the bound `resetAccounts` method.